src/static_canvas.class.ts

```diff
@@ -601,21 +620,26 @@ import { pick } from './util/misc/pick';
 
     /**
      * @private
-     * @param {fabric.Object} obj Object that was added
+     * @param {fabric.Object} object Object that was added
      */
-    _onObjectAdded: function(obj) {
-      this.stateful && obj.setupState();
-      if (obj.canvas && obj.canvas !== this) {
+    _onObjectAdded: function(object) {
+      this.stateful && object.setupState();
+      if (object.group && object.group instanceof fabric.ActiveSelection === false) {
         /* _DEV_MODE_START_ */
-        console.warn('fabric.Canvas: trying to add an object that belongs to a different canvas.\n' +
-          'Resulting to default behavior: removing object from previous canvas and adding to new canvas');
+        console.warn('fabric.Canvas: removing object from group before entering canvas', object, object.group, this);
         /* _DEV_MODE_END_ */
-        obj.canvas.remove(obj);
+        object.group.remove(object);
       }
-      obj._set('canvas', this);
-      obj.setCoords();
-      this.fire('object:added', { target: obj });
-      obj.fire('added', { target: this });
+      else if (object.canvas) {
+        /* _DEV_MODE_START_ */
+        console.warn('fabric.Canvas: removing object from canvas before entering another', object, object.canvas, this);
+        /* _DEV_MODE_END_ */
+        object.canvas.remove(object);
+      }
```

Comment:
in my opinion this should be executed always, not just with an else if.
What if it was in a group, on another canvas?
In that case the older canvas has a reference to an invalid object.
if we are safeguarding, we should remove from group and from canvas.

Reply:
> What if it was in a group, on another canvas?

I that case the first if block will catch it and `Group#remove` will clear refs.
Else if is necessary since we are checking for a parent and because an object always has a ref to canvas (even if not a direct child of canvas) we call tell if canvas is the parent or not only by checking if the `group` ref isn't empty.
I introduced `getParent` in a pending group PR for this purpose exactly.
Then it can become
```js
object.getParent()?.remove(object)

```